cpusupport/cpusupport_arm_crc32_64.c

```diff
@@ -0,0 +1,37 @@
+#include "cpusupport.h"
+
+#ifdef CPUSUPPORT_HWCAP_GETAUXVAL
+#include <sys/auxv.h>
+
+#if defined(__arm__)
+/**
+ * Workaround for a glibc bug: <bits/hwcap.h> contains a comment saying:
+ *     The following must match the kernel's <asm/hwcap.h>.
+ * However, it does not contain any of the HWCAP2_* entries from <asm/hwcap.h>.
+ */
+#ifndef HWCAP2_CRC32
+#include <asm/hwcap.h>
+#endif
+#endif /* __arm__ */
+#endif /* CPUSUPPORT_HWCAP_GETAUXVAL */
+
+CPUSUPPORT_FEATURE_DECL(arm, crc32_64)
+{
+	int supported = 0;
+#if defined(CPUSUPPORT_ARM_CRC32_64)
+#if defined(CPUSUPPORT_HWCAP_GETAUXVAL)
+	unsigned long capabilities;
+
+#if defined(__aarch64__)
+	capabilities = getauxval(AT_HWCAP);
+	supported = capabilities & HWCAP_CRC32;
+#elif defined(__arm__)
+	capabilities = getauxval(AT_HWCAP2);
+	supported = capabilities & HWCAP2_CRC32;
+#endif
+#endif /* CPUSUPPORT_HWCAP_GETAUXVAL */
+#endif /* CPUSUPPORT_ARM_CRC32_64 */
+
+	/* Return the supported status. */
+	return (supported > 0);
```

Comment:
Hmm, now that I'm looking at the other cpusupport files -- why not use the same layout, i.e.

```
#if defined(CPUSUPPORT_ARM_CRC32_64)
#if defined(CPUSUPPORT_HWCAP_GETAUXVAL)
unsigned long capabilities;

#if defined(__aarch64__)
capabilities = getauxval(AT_HWCAP);
return ((capabilities & HWCAP_CRC32) ? 1 : 0);
#elif defined(__arm__)
capabilities = getauxval(AT_HWCAP2);
return ((capabilities & HWCAP2_CRC32) ? 1 : );
#endif
#endif
#endif

return (0);
```

Reply:
In any case, I like the consistency of having these all `return ((foo) ? 1 : 0)` since that makes it immediately clear that the value is either 1 or 0.

Reply:
Would you be ok with
```
supported = (capabilities & HWCAP_CRC32) ? 1 : 0;
```

but retaining the single return statement?  (currently uploaded version)

The reason is pure banality: avoiding `clang -Weverything` complaining about unreachable code.  It would be possible to avoid this with various `#else \n return (0)` constructs, but that makes the code less readable.

The reason this isn't a problem for the other cpusupport files is that they have a single detection method (cpuid) which is valid for all relevant platforms.  In the case of ARM, we need to do platform-specific detection.  At the moment, we have linux ARM32 and linux AARCH64, but in a few months I expect to have FreeBSD (which may or may not have a similar split between 32 and 64-bit).  Having a single `int supported = 0; ... return (supported)`, and modifying the value of `supported` inside a relevant `#if #endif` construct, keeps this code simpler.

Reply:
Oh yeah, I forgot about the unreachable-code bit.